### ISSUE TITLE
Multi-treatment GRF

### ISSUE THREAD

**Issue by erikcs** [grf-labs/grf#733]
This is an overview issue - in the first round we aim for a causal forest with multiple categorical treatments `W` and a potentially vector valued outcome `Y`. Here is a table of the causal forest "taxonomy"

|    	| outcome Y   	| treatment W           	| instrument Z          	| name                         	|                               	|
|----	|-----	|-------------	|------------	|------------------------------	|-------------------------------	|
| a) 	| R   	| R           	|            	| `regression_forest`          	|                               	|
| b) 	| R  	| {1, ..., K}           	|            	| `probability_forest`          	|                               	|
| c) 	| R^K 	| R           	|            	| `multi_regression_forest`    	|                               	|
| d) 	| R   	| R / {0, 1}  	|            	| `causal_forest`              	| e(X): regression_forest       	|
| e) 	| R^K   	| {1, ..., K} 	|            	| `multi_arm_causal_forest` 	| e(X): probability_forest m(X): multi_regression_forest    	|
| f) 	| R   	| R^K         	|            	| `lm_forest`                          	| E(W|X): multi_regression_forest 	|
|    	|     	|             	|            	|                              	|                               	|
| g) 	| R   	| R / {0, 1}  	| R / {0, 1} 	| `instrumental_forest`        	|                               	|

_Building blocks:_

- [x] b) **Multi-class probability** estimation (`probability_forest`). #711 
    Future follow-ups:
    1. Incorporate sample weights into ProbabilitySplittingRule. #832 
    2. Tuning
    3. (Some potential performance modifications detailed in #711).
  <br/>
- [x] c) **Multi-outcome regression** (`multi_regression_forest`). #742 
    Can be used to estimate multivariate generalized propensity scores, and split on vector-valued pseudo outcomes.

    Future follow-ups:
    1. Tuning
    2. Variance estimates

_New forests:_

- [x] e) **Multi arm causal forest** (`multi_arm_causal_forest`). #748 
    Estimate treatment effects with multiple mutually exclusive categorical treatments 1,...K.
  
    Future follow-ups:
    1. Support for multivariate outcomes Y. #815 
    2. A `MultiRegressionSplittingRule` which supports more constraints, such as #W in each child. #921.
    3. Integrate with `best_linear_projection`
    4. Tuning
  <br/>

- [x] f) Some specific or generic forest-variant for multiple continuous treatments. #1138 

_General follow ups:_ 
- Performance enhancements to Data model, see issue #747
- [x] Update causal_benchmark.R. #831

(Thanks to @rajkumarkarthik for his original MGRF prototype in https://github.com/rajkumarkarthik/mgrf-develop)

**Comment by 0blongata** [comment]
I have been experimenting with multi_arm_causal_forest() over the last few days - very cool that the functionality to have multiple treatment groups was added.  I'm going through the example code here (https://grf-labs.github.io/grf/reference/multi_arm_causal_forest.html).

- I'm able to train mc.forest on the simulated example training data, everything works great.
- Now I'd like to test mc.forest on a separate test data set, which means estimating Y.hat to be able to get muA, muB, and muC.
(and estimating W.hat if an observational setting).
- It appears that mc.forest$Y.hat and mc.forest$W.hat are attached to the forest (mc.forest), and are estimated on the training OOB data.
- Is there a way to i. train mc.forest with training data, ii. use mc.forest to test a separate, test data set?
- I'm able to get tau_B and tau_C for the test data, but without a Y.hat estimate for the test data, I don't believe that I can back-calculate to get muA, muB, and muC.

Not sure if this is the right forum to post this, and maybe there is an updated policytree function that has the same functionality as multi_arm_causal_forest() ?

**Comment by erikcs** [comment]
If you want mu_W on a test set, you could train separate forests/estimators for Y.hat and W.hat, then use those to compute mu_W via predicted Y.hat, W.hat, using the same construction in that docstring (by default Y.hat is estimated with a regression forest(X, Y) and W.hat with a probability forest(X, W))

**Comment by 0blongata** [comment]
@erikcs  - thank you very much for the reply, I really appreciate it.

If I understood your response correctly, it’s suggested that three forests would be trained to estimate four different variables in the test data scenario.
1. mc.forest, trained on training data, used to estimate the tau_B_test and tau_C_test
2. mr.forest, trained on training data, used to estimate Y.hat_test
3. prob.forest, trained on training data, used to estimate W.hat_test

I think what I’m struggling with is that the goal of mc.forest is to estimate heterogeneous treatment effects (HTE). Functionally, mc.forest estimates tau_B and tau_C for the OOB training data, but also estimates Y.hat (and W.hat) for OOB data. mc.forest can be used with test data to estimate tau_B_test and tau_C_test, but not to estimate Y.hat_test or W.hat_test. If mc.forest can be used to estimate Y.hat (and W.hat) for OOB training samples, could it also be used to estimate Y.hat_test and W.hat_test for test data?

In my case, the motivation is to assess the performance of mc.forest for estimating HTE. In this context, it seems like evaluating mc.forest means evaluating its ability to estimate tau_B_test, and tau_C_test, but also Y.hat_test, in order to get mu_W_test. Though treatment propensity W.hat_test is also a required term to estimate mu_W_test in an observational setting, I see W.hat_test estimation as not directly part of estimating mc.forest’s performance, because W.hat_test estimation isn’t really part of the causal inference and can be addressed with “standard” ML approaches. However, maybe it’s also true that estimating Y.hat_test is also not part of causal inference - it’s just a prediction problem. Thus, like W.hat_test estimation, Y.hat_test estimation can also be addressed with “standard” ML approaches. However, to estimate tau_B_test and tau_C_test, which by definition is the E[Yi | Xi, Wi = B/C] - E[Yi | Xi, Wi = A], doesn't one have to estimate Y.hat_test? because of E[Yi | Xi, Wi = A] term?

One reason to estimate HTE would be if one was on the position to make treatment decisions. In that position, having mu_W would allow one to estimate the outcomes for all treatment options (including no treatment) and make an informed decision. If there are two different models that must be trained to estimate mu_W – a forest for tau_B and tau_C, and a forest for Y.hat – it seems like the both models would need to be evaluated for performance and also their statistical properties? However, maybe the mr.forest model doesn’t need to have certain statistical properties, because it’s only predicting and not being used for inference?

Note: The above could contain a number of misunderstandings on my part in terms of the details of how the algorithm works, as well as incorrect interpretations of estimands, etc.

Finally, the reason that to dive into this grf package, and specifically the multi_arm_causal_forest() function, is that I am working on a Masters thesis attempting to evaluate HTE estimation for multiple treatment arms and across multiple time points. For example, maybe there are four time points at which point the subject could receive treatment A, B, or C - so one could imagine each subject would have a vector of treatments c("B", "A", "C", "C"). I’m trying to understand multi_arm_causal_forest() before seeing if it could be applied to such a scenario. If it could, it might mean looking into extending the multi_arm_causal_forest() function for that scenario.

Below is an adaptation of the code in the docstring, trying to implement what you suggested:

## TRAIN
# Train a multi-arm causal forest with X, Y, and W
n <- 500
p <- 10
X <- matrix(rnorm(n * p), n, p)
W <- as.factor(sample(c("A", "B", "C"), n, replace = TRUE))
Y <- X[, 1] + X[, 2] * (W == "B") - 1.5 * X[, 2] * (W == "C") + rnorm(n)
mc.forest <- multi_arm_causal_forest(X, Y, W)
Y.hat <- mc.forest$Y.hat # estimations of Y.hat for OOB training data
W.hat <- mc.forest$W.hat # estimations of W.hat for OOB training data
W.hat <- matrix(data = 1/3, nrow = n, ncol = 3) # for simplicity we are in an RCT scenario, so can set W.hat = 1/3
mc.pred <- predict(mc.forest, drop = TRUE) #  use mc.forest to estimate tau_B, tau_C
tau.hat <- mc.pred$predictions
head(mc.forest$predictions, n = 25) == head(tau.hat, n = 25) # these are the same
# use tau_B, tau_C, Y.hat (and  W.hat if appropriate) to calculate mu_A, mu_B, mu_C
mu_A <- Y.hat - W.hat[,2] * tau.hat[, "B - A"] - W.hat[,3] * tau.hat[, "C - A"]
mu_B <- Y.hat + (1 - W.hat[,2]) * tau.hat[, "B - A"] - W.hat[,3] * tau.hat[, "C - A"]
mu_C <- Y.hat - W.hat[,2] * tau.hat[, "B - A"] + (1 - W.hat[,3]) * tau.hat[, "C - A"]

## TEST
# tau_B_test, tau_C_test estimate
# test m.forest with a new set of testing data.
n_test <- n
W_test <- W # keeping same balance of treatment groups
X_test <- matrix(rnorm(n_test * p), n_test, p) # different covariate than X
Y_test <- X_test[, 1] + X_test[, 2] * (W == "B") - 1.5 * X_test[, 2] * (W == "C") + rnorm(n_test) # generate test Y based on test X covariates
mc.pred.test <- predict(mc.forest, newdata = X_test, drop = TRUE) # gives tau_B, tau_C for X_test
tau.hat_test <- mc.pred.test$predictions
tau_B_test <- tau.hat_test[,1]
tau_C_test <- tau.hat_test[,2]

# Y.hat_test estimate
# training multi regression forest on X, Y to get Y.hat_test estimates
mr.forest <- multi_regression_forest(X = X, Y = Y) # train multi regression forest on training data to later predict Y.hat_test
Y.hat_test <- predict(mr.forest, newdata = X_test)
Y.hat_test <- Y.hat_test$predictions
# use tau_B_test, tau_C_test, Y.hat_test (and  W.hat_test if appropriate) to calculate mu_W_test
mu_A_test <- Y.hat_test - W.hat[,2] * tau.hat_test[, "B - A"] - W.hat[,3] * tau.hat_test[, "C - A"]
mu_B_test <- Y.hat_test + (1 - W.hat[,2]) * tau.hat_test[, "B - A"] - W.hat[,3] * tau.hat_test[, "C - A"]
mu_C_test <- Y.hat_test - W.hat[,2] * tau.hat_test[, "B - A"] + (1 - W.hat[,3]) * tau.hat_test[, "C - A"]

# mu_W_test should now be the estimated Yi value for the test data, given the treatment group Wi and the covariate matrix Xi.

**Comment by erikcs** [comment]
Hi @0blongata, W.hat and Y.hat are so-called nuisance components that are used in grf's approach to estimate tau(X), this [vignette](https://grf-labs.github.io/grf/articles/grf_guide.html) gives a brief overview of their role, and you are right evaluating those input estimates can be reduced to evaluating for ex predictive MSE. 

The code example to estimate mu_k(X) on Xtest would look like:
```R

library(grf)
n <- 500
p <- 10
X <- matrix(rnorm(n * p), n, p)
W <- as.factor(sample(c("A", "B", "C"), n, replace = TRUE))
Y <- X[, 1] + X[, 2] * (W == "B") - 1.5 * X[, 2] * (W == "C") + rnorm(n)

# Fit a Y.hat forest
Y.forest <- multi_regression_forest(X, Y)
Y.hat <- predict(Y.forest)$predictions
# Fit a W.hat forest
W.forest <- probability_forest(X, W)
W.hat <- predict(W.forest)$predictions

# Fit a MCF using precomputed Y.hat and W.hat
mcf <- multi_arm_causal_forest(X, Y, W, Y.hat = Y.hat, W.hat = W.hat)

# Predict mu_k(W) on X.test
ntest <- 100
Xtest <- matrix(rnorm(ntest * p), ntest, p)
tau.hat.test <- predict(mcf, Xtest, drop = TRUE)$predictions

# Construct mu_k(X) from tau_k(X)
Y.hat.test <- predict(Y.forest, Xtest)$predictions
W.hat.test <- predict(W.forest, Xtest)$predictions

Y.hat.baseline.test <- Y.hat.test - rowSums(W.hat.test[, -1, drop = FALSE] * tau.hat.test)
mu.hat <- cbind(Y.hat.baseline.test, c(Y.hat.baseline.test) + tau.hat.test)
colnames(mu.hat) <- levels(W)
```
